server/async_tcp.go

```diff
@@ -95,15 +97,44 @@ func (s *AsyncServer) FindPortAndBind() (err error) {
 	})
 }
 
-// Run starts the server, accepts connections, and handles client requests
-func (s *AsyncServer) Run(ctx context.Context) error {
-	defer func(fd int) {
-		err := syscall.Close(fd)
-		if err != nil {
+// ClosePort ensures the server socket is closed properly.
+func (s *AsyncServer) ClosePort() {
+	if s.serverFD != 0 {
+		if err := syscall.Close(s.serverFD); err != nil {
 			log.Warn("failed to close server socket", "error", err)
+		} else {
+			log.Info("Server socket closed successfully")
 		}
-	}(s.serverFD)
+		s.serverFD = 0
+	}
+}
+
+// WaitForSignal listens for OS signals and triggers shutdown
+func (s *AsyncServer) WaitForSignal(cancel context.CancelFunc, sigs chan os.Signal) {
```

Comment:
IMO this may not be needed if we ensure InitiateShutdown() is called before the context is cancelled.

Reply:
Makes sense, fixed and verified.